rpc/src/v1/impls/eth.rs

```diff
@@ -451,11 +456,38 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> EthClient<C, SN, S
 				self.miner
 					.pending_state(info.best_block_number)
 					.map(|s| Box::new(s) as Box<dyn StateInfo>)
-					.unwrap_or(Box::new(self.client.latest_state()) as Box<dyn StateInfo>)
+					.unwrap_or_else(|| {
+						// todo[dvdplm]: this strikes me as odd. If I call `get_state` with
+						// `BlockNumber::Latest` I get a `StateOrBlock::Block` back, but if I call
+						// it with `BlockNumber::Pending` and the state is not there so the fallback
+						// is triggered, then I get a `StateOrBlock::State` back. Shouldn't it be
+						// consistent?
+						warn!("Asked for best pending state, but none found. Falling back to latest state");
+						Box::new(self.client.latest_state()) as Box<dyn StateInfo>
+					})
 					.into()
 			}
 		}
 	}
+
+	/// Get the state and header of best pending block. On failure, fall back to the best imported
+	/// blocks state&header.
+	// todo[dvdplm]: better name?
+	// todo[dvdplm]: Scary thought: what if the state is not pending but the header is? We'd return the state from
```

Comment:
That's a good point. We use `best_block_number` here, which feels a bit incorrect as the block may be imported in the meantime. Would be way better to get the `hash` of latest block and consistently use `hash` to extract the information we need.